modules/p2/ding_interests/ding_interests.module

```diff
@@ -0,0 +1,652 @@
+      return $term;
+    }
+  }
+
+  // Return any.
+  return array_values($terms)[0];
+}
+
+
+/**
+ * Implements hook_theme().
+ */
+function ding_interests_theme($existing, $type, $theme, $path) {
+  $theme = array(
+    'ding_interests_tree' => array(
+      // 'template' => 'ding-in',
+      'variables' => array('root' => 0, 'tree' => array()),
+      // 'path' => $path . '/theme',
+    ),
+  );
+  return $theme;
+}
+
+/**
+ * Preprocess function for Ding interests: Overview.
+ */
+function template_preprocess_ding_interests_tree(&$variables) {
+  $voc = taxonomy_vocabulary_machine_name_load('ding_content_tags');
+  $tree = taxonomy_get_tree($voc->vid, $variables['root'], 2, TRUE);
+  $actual_tree = array();
+
+  if ($variables['root'] !== 0) {
+    $root = taxonomy_term_load($variables['root']);
+    $root->parents = array('0');
+    $tree = array_merge(array($root), $tree);
+  }
+
+  $count = count($tree);
+  if ($count == 1) {
+    return;
+  }
+
+  $access = ding_interests_get_status_tids();
+  for ($i = 0; $i < $count; $i++) {
+    if (ding_base_get_value('taxonomy_term', $tree[$i], 'field_status', 'tid') == $access['denied']) {
+      array_splice($tree, $i, 1);
+      $i -= 1;
+      $count -= 1;
+    }
+    else {
+      if ($tree[$i]->parents[0] == '0') {
+        $actual_tree[$tree[$i]->tid] = $tree[$i];
+      }
+      else {
+        if (isset($actual_tree[$tree[$i]->parents[0]])) {
+          $actual_tree[$tree[$i]->parents[0]]->children[$tree[$i]->tid] = $tree[$i];
+        }
+      }
+    }
+  }
+
+  $variables['tree'] = $actual_tree;
+}
+
+/**
+ * Todo: doc here.
+ */
+function theme_ding_interests_tree($variables) {
+  $parent = 0;
+  $out = '';
+  $count = count($variables['tree']);
+  if ($count == 0) {
+    return $out;
+  }
+
+  foreach ($variables['tree'] as $item) {
+    $uri = entity_uri('taxonomy_term', $item);
+
+    $out .= '<div class="item">';
+    if (arg(0) != 'taxonomy') {
+      $out .= '<h2>' . l($item->name, $uri['path']) . '</h2>';
+    }
+    else {
+      $out .= '<h2>' . t('Sub subjects for !link', array('!link' => l($item->name, $uri['path']))) . '</h2>';
+    }
+    if (isset($item->children)) {
+      $items = array();
+      foreach ($item->children as $child) {
+        $uri = entity_uri('taxonomy_term', $child);
+        $items[] = l($child->name, $uri['path']);
+      }
+
+      $out .= theme('item_list', array('items' => $items));
+    }
+    $out .= '</div>';
+  }
+
+  return $out;
+}
+
+/**
+ * Todo: doc here.
+ */
+function ding_interests_ctools_plugin_directory($owner, $plugin_type) {
+  if ($owner == 'ctools' && $plugin_type == 'content_types') {
+    return 'plugins/content_types';
+  }
+}
+
+/**
+ * Get the status granted and denied tids.
+ */
+function ding_interests_get_status_tids() {
+  // Get access granted or denied term tids.
+  $vocab = taxonomy_vocabulary_machine_name_load('interest_status');
+  $tree = taxonomy_get_tree($vocab->vid);
```

Comment:
Why do we need to load the vocabulary and tree? We already have the tids in the variables so could it be reduced to this:

```php
return array(
  'denied' => variable_get('ding_interests_optedout_tid'),
  'granted' => variable_get('ding_interests_selected_tid')
);
```